components/calendar/src/chinese_based.rs

```diff
@@ -309,6 +337,159 @@ impl<C: ChineseBased + CalendarArithmetic> ChineseBasedDateInner<C> {
         debug_assert!(result < MAX_ITERS_FOR_MONTHS_OF_YEAR, "The given year was not a leap year and an unexpected number of iterations occurred searching for a leap month.");
         result
     }
+
+    /// Create a new arithmetic date from a year, month ordinal, and day with bounds checking; returns the
+    /// result of creating this arithmetic date, as well as a ChineseBasedCache - either the one passed in
+    /// optionally as an argument, or a new ChineseBasedCache for the given year, month, and day args.
+    pub(crate) fn new_from_ordinals(
+        year: i32,
+        month: u8,
+        day: u8,
+        cache_arg: Option<ChineseBasedCache>,
+    ) -> (Result<ArithmeticDate<C>, CalendarError>, ChineseBasedCache) {
+        let (max_month, cache) = Self::months_in_year_cached(year, cache_arg);
+        if month > max_month {
+            return (
+                Err(CalendarError::Overflow {
+                    field: "month",
+                    max: max_month as usize,
+                }),
+                cache,
+            );
+        }
+
+        let max_day_tuple = Self::days_in_month_cached(year, month, Some(cache));
+        let max_day = max_day_tuple.0;
+        if day > max_day {
+            return (
+                Err(CalendarError::Overflow {
+                    field: "day",
+                    max: max_day as usize,
+                }),
+                cache,
+            );
+        }
+
+        // Unchecked can be used because month and day are already checked in this fn
+        (
+            Ok(ArithmeticDate::<C>::new_unchecked(year, month, day)),
+            cache,
+        )
+    }
+
+    /// Call `months_in_year_cached` on a `ChineseBasedDateInner`
+    pub(crate) fn months_in_year_inner(&self) -> u8 {
+        Self::months_in_year_cached(self.0.year, Some(self.1)).0
+    }
+
+    /// Return the number of months in a given year, which is 13 in a leap year, and 12 in a common year.
+    /// Also returns a ChineseBasedCache - either the cache passed as an argument, if applicable, or a new
+    /// ChineseBasedCache for the given year.
+    fn months_in_year_cached(
+        year: i32,
+        cache_arg: Option<ChineseBasedCache>,
```

Comment:
praise: nice work on the cache -> cache functions. we could have done stuff mutably but i think that makes it more complicated here